### PR TITLE
emscripten: 2.0.1 -> 2.0.25

### PR DESCRIPTION
**Pull request by khovansky-al**
###### Motivation for this change
Taking advantage of latest SIMD spec in WASM requires a much newer version of emscripten as well as (yet unreleased) clang 13 and a newer binaryen.

This PR updates emscripten to v2.0.25 (latest tag release). Binaryen and llvm were updated to revisions specified in [emscripten releases repo](https://chromium.googlesource.com/emscripten-releases/) DEPS file. See [emscripten-releases-tags](https://github.com/emscripten-core/emsdk/blob/main/emscripten-releases-tags.txt) for the mapping of tags in `emscripten-releases` to emscripten versions.
I imagine `llvm_git` could be updated any moment which could at some point break emscripten. Not sure what would be the best course of action, as `llvm_git` doesn't seem to allow overrides.

Starting with clang 12, one of LLVM's modules is miscompiled under gcc resulting in random segfaults when precompiling emscripten libraries cache. I added a workaround taken from the bug's page: https://bugs.llvm.org/show_bug.cgi?id=50611

As for testing, out of `emscriptenPackages` on current master only `zlib` appears to be building. `libxml2`, `xmlmirror` and `json_c` are already broken.
I made adjustments that keep `zlib` building as well as fix `libxml2` and `json_c`. `xmlmirror` appears to be more deeply broken and I don't have time to fix it.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by khovansky-al** [comment]
Just realized `llvmPackages_git` is not supposed to be used for nixpkgs. Well, guess this is not going to work.

File: pkgs/development/compilers/binaryen/default.nix [hunk 1]
```diff
@@ -2,23 +2,15 @@
...
     owner = "WebAssembly";
     repo = "binaryen";
     rev = "version_${version}";
-    sha256 = "1mqpb6yy87aifpbcy0lczi3bp6kddrwi6d0g6lrhjrdxx2kvbdag";
+    sha256 = "15gvfai3snndlgsppyjjf17xw4bmyhwm2fk07wsvfd9ahi09kn5c";
   };
 
-  patches = [
-    # Adds --minimize-wasm-changes option required by emscripten 2.0.1
-    (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/binaryen/default.nix:16]: Please remove the unused fetchpatch from inputs.

File: pkgs/development/compilers/emscripten/default.nix [hunk 2]
```diff
@@ -26,21 +26,40 @@ stdenv.mkDerivation rec {
...
     repo = "emscripten";
-    sha256 = "06dsd819qjv4n2ihrz1mpn5aigmbv0gpkm7iw06wrqx30nzphnpk";
+    sha256 = "0kmfvd6h0p7a95z6j941vx7bcspfbxfja5ab0qihhzizfzn6719b";
     rev = version;
   };
 
   nativeBuildInputs = [ makeWrapper ];
   buildInputs = [ nodejs python3 ];
 
+  resourceDir = "${llvmEnv}/lib/clang/13.0.0/";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/emscripten/default.nix:36]: Please substitute the version number.

File: pkgs/development/compilers/emscripten/default.nix [hunk 3]
```diff
@@ -26,21 +26,40 @@ stdenv.mkDerivation rec {
...
+    # invocations directly.
+    #
+    # Important note: with non-nix clang, sysroot/include dir ends up being the last
+    # in the include search order, right after the resource root.
+    # Hence usage of -idirafter. Clang also documents an -isystem-after flag
+    # but it doesn't appear to work
+    substituteInPlace emcc.py \
+      --replace \
+        "['--sysroot=' + shared.Cache.get_sysroot(absolute=True)]" \
+        "['--sysroot=' + shared.Cache.get_sysroot(absolute=True), '-resource-dir=${resourceDir}', '-idirafter' + shared.Cache.get_sysroot(absolute=True) + os.path.join('/include'), '-iwithsysroot' + os.path.join('/include','c++','v1')]"
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/emscripten/default.nix:53]: Maybe we should use a patch instead.

**Comment by yu-re-ka** [comment]
I have a seemingly working emscripten 2.0.26 on my local branch, should I push it to this PR or create a new one @khovansky-al ?

**Comment by khovansky-al** [comment]
@yu-re-ka thanks for taking on this. Feel free to open a new PR, I'll close this one in favor of yours. A bit busy with other things these days and probably won't have time to sit down and properly finish this for another week or two.

P.S. appreciate the attribution in your branch's commits. For Github to pick it up though I think the trailer should be `Co-authored-by: ...`.

**Comment by yu-re-ka** [comment]
https://github.com/NixOS/nixpkgs/pull/133217/commits

**Pull request closed without merging**